### DISCUSSION TITLE
Understanding How next.js works under the hood ?

### DISCUSSION THREAD

**Discussion by tanvirraj** [vercel/next.js#11826] [Help]
Let's say I'm a beginner who uses next.js. 
if I want to understand how next.js works under the hood and what every feature works the next.js offer where I can start?
I want to dig deeper and in future intention to contribute the codebase 

can anyone introduce me where I began ?

**Comment by puntope** [comment]
You can start like this -> `npm i react react-dom next`

Just kidding,

If you're a beginner I think the first step is to be an advanced user, that means:

1. Read well the public docs https://nextjs.org/docs/getting-started
2. Build projects using NextJS, play with it.
3. Ask and participate in the community 
4. Read the public code in this repo

Kindest

**Reply by tanvirraj** [reply]
Thanks for your replay. 
Actually I build a small website with nextjs. 
but I want to go deeper into the code

**Reply by SeckMohameth** [reply]
Same!

**Comment by littlepoolshark** [comment]
I am also curious too...but why is this issue so little concerned?

**Reply by jarmos-san** [reply]
More often than not, people are lazy & wants to get work done for the day so they can get back to their hobbies, family & resting time (not blaming them for being lazy or anything though). :sweat_smile:

**Comment by jarmos-san** [comment]
The recipe to learn how any technology works under-the-hood is all the same. With that in mind, here's what I would do:
1. As mentioned in a previous comment, I would use Next.js enough to become "_an advanced user_" first. By that I mean, having a proper understanding of what `getStaticProps`, `getServerSideProps` & Incremental Static Regeneration (ISR) as well.
2. And then I would fork the repository to take a better look at Next.js's source code to get a general idea of it's structure & organisation.
3. Once I'm comfortable with the source code, I would attempt to an issue or two every week.

In other words, there's no shortcut to learn about the inner workings of Next.js. You use it enough first & then contribute to the project, it's that simple!

Also here are some resources I stumbled across while trying to understand Next.js's inner workings myself:

- [The Next.js Handbook | freeCodeCamp](https://www.freecodecamp.org/news/the-next-js-handbook/#view-source-to-confirm-ssr-is-working) by Flavio Copes (**the contents are a bit outdated**, for e.g it doesn't shed light on ISR & discusses the new deprecated `getInitialProps` in detail).
- The [official Next.js tutorial](https://nextjs.org/learn/foundations/about-nextjs) which isn't as comprehensive as I would expect it to be but it touches upon some of the most important topics of the framework.
- [Next.js: The Good, Bad and Ugly](https://dhanrajsp.me/blog/the-troubles-with-nextjs) by Dhanraj Padmashali which is the **MOST** comprehensive write-up I've come across about the topic till now!
- [Next.js Performance: Making a Fast Framework Even Faster](https://calibreapp.com/blog/nextjs-performance) is a good write-up which discusses some best-practices to develop performant websites using the framework.

So, the resources I shared above are only a handful & I believe there are way more hidden gems out there! But regardless, my suggestions would be to keep using Next.js for your projects. Identify some pain points & then try to contribute to the project for the best learning experience! And yes, don't forget to share you experience & learning about Next.js's inner workings later on either.

**Reply by Fabiote2003** [reply]
I really appreciate your comment! Thanks!

**Comment by itzsrikanth** [comment]
Actually I am on the same path as well. I have learned some of the inner workings of Glimmer.Js for a HTML minification project (part of Ember.Js) this way. Hoping to repeat the same with Next.Js. This method has its drawback, but it was getting hard to find many implementations of recent version of Glimmer.Js in the wild, unlike React based community.

I started with adding breakpoints in VSCode and stepping into the modules to know the overall flow of logic. Then started to go through the tests for each module to understand the requirements. Once we know the input and output of each module, a familiarity starts to emerge from the code. It is comparatively easier to understand the inner workings now. All we need to do is put together the modules piece by piece.

Now repeat the entire step multiple times. Imagine yourself as an multi-stage parser and share your learnings with us :)

**Comment by jancbeck** [comment]
Back in the days I learned a lot from annotated sources (see [jquery](https://underscorejs.org/docs/underscore-esm.html), [underscore](https://underscorejs.org/docs/underscore-esm.html), [backbone](https://backbonejs.org/docs/backbone.html)). 

These would just pull out the comments from the source and display side by side. 

Unfortunately, today’s modular structure of projects and tooling makes this harder to accomplish I think. 

**Edit**: just occurred to me that a LLM might be a great tutor for digging deeper into a framework like Next. It can abstract away and get to the essentials of how things work. 
I created a [custom GPT](https://chat.openai.com/g/g-ICC0n8PjU-next-js-architecture-tutor) for this purpose and loaded the entire source of Next 14.0.3 as a zip into it but that's unlikely to work since that's about 21000 individual files. Same if I open the source in VS Code and use Github Copilot. Current context windows are too small at current generation LLMs. It tends to give outdated answers based on Next 13 code which it probably got from its initial training run or it just hallucinates or makes up pseudo-code answers. When I asked it how Next processes requests for static files it confused that with static site exports instead.

Better ask it more general questions and ask it for the name of the class of function that implements a certain feature. Then browse that code yourself. If something isn't clear, highlight it in VS Code and let Copilot explain what's going on.

**Comment by daniele-roncaglioni** [comment]
Hi all,

a while ago I had the same question so I made a diagram for myself showing:
- how project files get built into what is then actually used
- which files get built in which cases, how, and the usage
- the request flow...how are these files fetched/used by the browser in different cases

I uploaded it to [this repo](https://github.com/daniele-roncaglioni/nextjs-under-the-hood).

Since I made this a while ago I only have it for the Page Router. Would be happy if somebody wants to help do the same for the App Router or in general contribute content, otherwise I'll just add that myself in some time.

**Comment by HarimbolaSantatra** [comment]
There's a page about [Architecutre](https://nextjs.org/docs/architecture) now on NextJS doc.
